**CHI SQUARE DISTRIBUTION:**

this is used for discrete data.
--> here we calculate the proportion, not the mean.
--> In ANOVA, we use mean.

**1) Goodness of fit - one variable**

**2) Degree of Association - two variables**



**1)GOODNESS OF FIT**

In [ ]:
import numpy as np
import pandas as pd
import scipy
from scipy import stats
from scipy.stats import chisquare

In [ ]:
pattern = np.array([0.6,0.25,0.1,0.05])

In [ ]:
sum(pattern)

1.0

In [ ]:
obs_counts = np.array([250,180,120,40])

**H0:** **Counts from the Sample follow the pattern**

In [ ]:
sum(obs_counts)

590

In [ ]:
exp_counts = 590*pattern
exp_counts

array([354. , 147.5,  59. ,  29.5])

**the expected counts are different from the observed counts**

In [ ]:
scipy.stats.chisquare(obs_counts,exp_counts)

# this is the main result.
# as test statistic increses, pvalue decreases
#

Power_divergenceResult(statistic=104.51977401129943, pvalue=1.657508756495914e-22)

--> p value < 0.05

--> the null hypothesis is rejected

--> the counts do not follow the pattern

--> there is some actual difference apart from chance variation.

**Why do we do all these stats that we done until now when we do the visual inspection and tell ?**

--> 1) To check whether the cjhance variation is present or not?
--> 2) practical significane is calculated. this can chnage from person to person and situation to situation.

In [ ]:
d = obs_counts - exp_counts
d

#deviation between observed values and expected values

array([-104. ,   32.5,   61. ,   10.5])

In [ ]:
c = d**2/exp_counts
c

#63.06 is more contributing here.
#Statistcally, the difference b/w the 59 and 120 is significant.
#python does not show the breakup values(this array) to us. That's why we're calculating manually

array([30.55367232,  7.16101695, 63.06779661,  3.73728814])

In [ ]:
sum(c)

#formula what we did is sum of ((observed counts - expected counts)**2)/Expected counts
#if observed counts = expected counts, test statistic= 0
#if test statistic !=0, then there is some samples which are playing a significant role in the difference(see the next text box)
#this value is equal to the test statistics value in chi square command.

104.51977401129943

**Statistcally, the difference b/w the 59 and 120 is significant.**

2nd test under CHI SQUARE FAMILY:


**2) DEGREE OF ASSOCIATION:**

--> here we have two variables

--> each of these two variables, they have multiple categories eg: gender, different branches

supose we're creating a table, we write columns has medical, science etc
in rows only 2, either person gets a job or does not get a job.

**what we're going to check?**
--> is there a association with your qualification and kind of job you get? This is called degree of association.

Ex:
45 - btech got the job
27 - bsc
21 - b com


[31,28,27]
31 - btech students who did not get the job
28 - bsc
27 - b com

H0: Whether you get the job or not does not depend on the qualification



In [ ]:
a1 = np.array([[45,27,21],[31,28,27]])
a1

array([[45, 27, 21],
       [31, 28, 27]])

In [ ]:
from scipy.stats import chi2_contingency

In [ ]:
scipy.stats.chi2_contingency(a1)

Chi2ContingencyResult(statistic=3.0780934692405153, pvalue=0.2145855609456467, dof=2, expected_freq=array([[39.48603352, 28.57541899, 24.93854749],
       [36.51396648, 26.42458101, 23.06145251]]))

statistic=3.0780934692405153,
pvalue=0.2145855609456467
p>0.05
hence H0 is not rejected
hence, you ccan choose any branch, job does not depend on the quaification.
no change is required.

dof = (no of rows -1) *( no of col -1)

observed values:
array([[45, 27, 21],
       [31, 28, 27]])

expected values:
expected_freq=array([[39.48603352, 28.57541899, 24.93854749],
       [36.51396648, 26.42458101, 23.06145251]]))

--> The differnce between the observed values and expected values is not significant because p > 0.05

**MACHINE LEARNING:**

--> We give some data, machine makes some pattern and predict the output.

--> Unless and until the relevant data is given, we cannot make the prediction.

--> there must be some conditions under which the data has been collected. because while prediction, if the conditions change then the data will be useless. The predi

eg: insecond wave, delta variant data is collected, but in third wave, omicron has came. so the condition has changed. the delta variant data is useless for predicting the omicron virus effect.


**1) Supervised learning:**

--> In this we use **Markup data.**
--> Here we have a dedicted column for the output so that we can create the training and the testing models.
--> in our module, we have 80% supervising learning as **Regression**

**2) Unsupervised learning:**



in unsupervised, we have 2 models:
**i) clustering**
**ii) Dimensionality reduction**



The first algorithm in supervised learning:



**SIMPLE LINEAR REGRESSION:**

In case of SIMPLE LINEAR REGRESSION, the Response is Continuous, and there will be Single predictor which is also continuous

The output for SLR will be in the form of ** y = mx+c**, where m is the slope, c is y intercept, we're finding the values of m and c

if there are two factors, then The outtput for SLR will be in the form of **y = m1x1+m2x2+c** , where m is the slope, c is y intercet.

**Correlation coefficient(r):**

This gives the information about the relationship of two continuous variables.
range of r : -1 to +1

This has two parts:
**i) Sign:**

The sign of Correlation coefficient tells us type of relationship
**this can be +ve --> positive relationship:**
This means if we increase the  value of first variable (x), if the second variable (y) value increases then we call +ve relationship.

eg: age of car and expenses of the servicing


**this can be -ve --> negative relationship**

This means if we increase the  value of first variable, if the second variable value decreases then we call -`ve relationship

ex: age of car and mileage of car
age of car and resale value of car.

**ii) Magnitude:**
This tells the strength  of the relationship.
The closer the magnitude gets to 1 or -1, stronger is the relationship betwen the two variables, the strength of the relation increases.

ex: to predict a, we can use either b or c. you can choose only one.

rAB = 0.87
rAC = -0.96

which is used for making predictions?

ans: we have to use -0.96. while choosing the strength, ignore the sign and choose the highest value closer to 1.

hence we choose rAC.

ex2:
the r between a and b is 0.1 or 0 can we say, there is no relationship?

**r =0 indicates,there is no linear relation between a and b, but non-linear relationship can be present.**
the output will no come y=mx+c, but may be different.


**When the process is unstable, don't make predictions.**

**ex: the first 3-4 months drving of a new bike, the mileage is not stable, so we can't make predictions.**

--> In the context of simple linear regression (SLR), the null hypothesis typically refers to the hypothesis that there is no significant relationship between the predictor variable and the response variable.

-->More specifically, the null hypothesis states that the true slope coefficient (m) in the population is equal to zero, indicating no linear relationship between the predictor and the response.

The null hypothesis for SLR can be stated as:

H₀: m = 0

This hypothesis assumes that the predictor variable has no effect on the response variable, or any observed relationship is due to chance or random variation.

To test the null hypothesis, statistical techniques are used, such as hypothesis testing or t-tests. These tests assess the significance of the estimated slope coefficient (m) and determine whether it is significantly different from zero. If the null hypothesis is rejected, it suggests evidence of a significant linear relationship between the predictor and the response variable.

In [ ]:
import numpy as np
import pandas as pd
import statsmodels
import sklearn
import statsmodels.api as sm
from statsmodels.formula.api import ols
import statsmodels.stats.multicomp
import sklearn
from sklearn.model_selection import train_test_split
from sklearn.metrics import confusion_matrix
from sklearn.metrics import classification_report


A linear regression model establishes the relation between a dependent variable(y) and at least one independent variable(x)

statsmodels : provides classes and functions for the estimation of many different statistical models.

pandas : library used for data manipulation and analysis.

NumPy : core library for array computing

In [ ]:
import os

In [ ]:
os.getcwd()

'/content'

In [ ]:
os.chdir(r'/content/drive/MyDrive/STATSCDAC')

In [ ]:
os.getcwd()

'/content/drive/MyDrive/STATSCDAC'

In [ ]:
df= pd.read_excel('CDAC_DataBook.xlsx', sheet_name='faithful')

In [ ]:
df.head()
#eruptions tlls us how long the eruptions takes place.
#waiting : tells us how much time the eruprtions takes place

eruptions  waiting
0      3.600       79
1      1.800       54
2      3.333       74
3      2.283       62
4      4.533       85

See your machine learning notes in your notebook.

In [ ]:
x_train, x_test, y_train, y_test = train_test_split(df['waiting'],df['eruptions'], test_size=0.25)

# Segregating the dataset into training and testing datasets.
# 25% testing data set, 75% training dataset
# x_train, x_test are predictors
# y_train. y_test are response

In [ ]:
x_train = sm.add_constant(x_train,prepend=False)

#now the model is of the form y=mx+c
#we're giving possibility to python that, if it's want to add constant.

In [ ]:
x_train.head()

200    60
190    81
139    79
239    64
193    84
Name: waiting, dtype: int64

In [ ]:
y_train.head()

200    2.100
190    4.800
139    3.733
239    2.333
193    4.100
Name: eruptions, dtype: float64

**Creating a model:**

**Ordinary Least Squares regression (OLS):**

Ordinary Least Squares regression (OLS) is a common technique for estimating coefficients of linear regression equations which describe the relationship between one or more independent quantitative variables and a dependent variable (simple or multiple linear regression)

In [ ]:
mod1 = sm.OLS(y_train,x_train).fit()

#creating a model
#OLS =Ordinary Least squares.
# performing the regression
# and fitting the model
#the .fit() method will calculate the mean (µ) and the standard deviation (σ) of the particular feature F. We can use these parameters
# later for analysis.

--> sm refers to the statsmodels library, which provides a wide range of statistical models and tools for data analysis.

--> sm.OLS creates an ordinary least squares (OLS) regression model object.

--> The fit() method is then called on the model object to estimate the coefficients of the linear regression model using the training data.

-->The fit() method returns a fitted model object that contains various attributes and methods for analyzing the results of the regression.

In [ ]:
print(mod1.summary())

                            OLS Regression Results                            
Dep. Variable:              eruptions   R-squared:                       0.798
Model:                            OLS   Adj. R-squared:                  0.797
Method:                 Least Squares   F-statistic:                     798.8
Date:                Tue, 20 Jun 2023   Prob (F-statistic):           4.02e-72
Time:                        06:17:45   Log-Likelihood:                -150.52
No. Observations:                 204   AIC:                             305.0
Df Residuals:                     202   BIC:                             311.7
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
waiting        0.0758      0.003     28.262      0.0

Summary:

**Step-1:**

in our case x is predictor
we want output of m and c --> y=mx+c
it gives the values of m and c
for both values we have null hypothesis for each.
H0 for m H0 for c

H0 for m:
m = 0  which means we can give any value for x. it will be multiplied by 0. x has no effect on y.

H0 for c:
c = 0

if no of variables =1, no of H0 = 2
if no of variables =3, no of H0 = 4 (x1,x2,x3,c)

**Analysis of summary:**

**1) coef of waiting = 0.0758 is the value of m**

**2) coeff of  const = -1.8965 is the value of c**

**y=mx+c**
**y=0.0758 x + (-1.8965) **

bare minimum is that You should be able to tell the equation of the model in the interview.

**Step -2:**
--> Prob (F-statistic): 4.02e-72

--> As p value is <0.05, H0 is rejected.

--> The influence of x on y is not a matter of chance variation, it cannot be ignored, it is statistically significant.
--> The influence of c on y is not a matter of chance variation, it cannot be ignored, it is statistically significant.

**Step - 3**:

**coefficient of determination:**:

--> R-squared: 0.798.
--> Finding the R and squaring it.

--> only in the case of single predictor and single response we can say the coefficient of determination= correlation coeffient. R square is 0.798 in this case.

--> in other cases it may not be the case because of morethan 1 predictor.

--> the model can explain 80 % of the variation, 20% of the variation it can't explain.


**The difference between the predicted value and actual result is called Residual.**

Now take this values and tell the response:


In [ ]:
x_test = sm.add_constant(x_test, prepend = False)

In [ ]:
y_pred = mod1.predict(x_test)

In [ ]:
y_pred[:5]

#These are the values predicted by the model. compare these values with the expected values which we already know. (see below)

53     4.172956
27     3.869646
39     4.931231
36     1.746477
132    2.353097
dtype: float64

In [ ]:
y_test[:5]

#these are the results we already know.

53     4.833
27     4.083
39     4.783
36     1.867
132    2.800
Name: eruptions, dtype: float64

Another example:
changing the dataset

**MULTIPLE LINEAR REGRESSION:**

In [ ]:
df= pd.read_excel('CDAC_DataBook.xlsx', sheet_name='stackloss')
df.head()

AirFlow  WaterTemp  AcidConc  StackLoss
0       80         27        89         42
1       80         27        88         37
2       75         25        90         37
3       62         24        87         28
4       62         22        87         18

airflow,watertemp, acidconc are predictors, stackloss is the response

In [ ]:
x_train, x_test, y_train, y_test = train_test_split(df.drop('StackLoss',axis=1),df['StackLoss'], test_size=0.25)

In [ ]:
x_train = sm.add_constant(x_train,prepend=False)

In [ ]:
mod1 = sm.OLS(y_train,x_train).fit()
mod1

In [ ]:
print(mod1.summary())

                            OLS Regression Results                            
Dep. Variable:              StackLoss   R-squared:                       0.898
Model:                            OLS   Adj. R-squared:                  0.870
Method:                 Least Squares   F-statistic:                     32.24
Date:                Tue, 20 Jun 2023   Prob (F-statistic):           9.56e-06
Time:                        07:13:50   Log-Likelihood:                -39.154
No. Observations:                  15   AIC:                             86.31
Df Residuals:                      11   BIC:                             89.14
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
AirFlow        0.6681      0.186      3.583      0.0

/usr/local/lib/python3.10/dist-packages/scipy/stats/_stats_py.py:1736: UserWarning: kurtosistest only valid for n>=20 ... continuing anyway, n=15
  warnings.warn("kurtosistest only valid for n>=20 ... continuing "


StackLoss= 0.6681 AirFlow + 1.4136 WaterTemp  - 0.1772 AcidConc  -36.9953

H0 for acid conc: acid conc has no effect on Stackloss
H0 for airflow : airflow has no effect on Stackloss
H0 for watertemp : watertemp has no effect on Stackloss

H0 for acid conc is not rejected as p>0.05
acid conc has no effect on Stackloss.

H0 for airflow is rejected as p < 0.05
airflow has an effect on stackloss.
airflow's effect is statistically significant on stackloss

H0 for watertemp is rejected as p < 0.05
watertemp has an effect on stackloss.
watertemp's effect is statistically significant on stackloss

**Coefficient of Determination(R-squared):**

R squared : 0.898 is closer to 1. This is a better model.


**Adavnced explanation:**
**Just for indepth understanding:**

Here the Acid Conc has some effect on Stackloss, but with respect to airflow and WaterTemp it is not significant and can be ignored.

The sign of the coeff tells you whether it has the positive relation or negative reation.
here the stack loss has  a positive relation with Airflow, positive relation with watertemp, negative relation with acidconc.


--> what the coeff tells us?

in case of airflow, if we keep other values constant and increase the airflow by 1 unit, then the Stackloss increases by 0.6681 units.
--> SL = 3+0.6*AF
       = 3+0.6(AF+1)
       = 3+0.6AF + 0.6

--> **which factor's p value is low, then it's effect is biggest.**
for checking the impact, we check the p-value, not the coefficient.
here, airflow has the biggest impact among the factors.

--> If two p-values are same, then check for t values to find the impact.
**which factor's t value is low, then it's effect is biggest**

**t value = coeff/std error**

where **std error = std dev/ sqrt(n)**


--> Lets assume all the three null hypothesis are not rejected.
then none of the factor has impact or influence on the response or output.

**GENERAL REGRESSION/CATEGORICAL REGRESSION:**

-->**One or more of the predictors can be categorical data**.

--> **Response is continuous**.

when we work with categorical data, we need to introduce dummy variables.

lets take an example,

exp and quali are predictors, salary is response.
in qualification we have engineer, normal sci grasduate, commerce graduate

Experience (coni data):    
Quali(categorical data):   
Salary(response):

in quali, python will take the first category in alphabetical order ,in this case it is commerce.

E AND S are introduced:

refer notes for table:








In [ ]:
import numpy as np
import pandas as pd
import statsmodels
import sklearn
import statsmodels.api as sm
from statsmodels.formula.api import ols
import statsmodels.stats.multicomp
import sklearn
from sklearn.model_selection import train_test_split
from sklearn.metrics import confusion_matrix
from sklearn.metrics import classification_report

In [ ]:
os.chdir(r'/content/drive/MyDrive/STATSCDAC')

In [ ]:
df = pd.read_excel('CDAC_DataBook.xlsx',sheet_name='salaries')

In [ ]:
df.head()

rank discipline  yrs_phd  yrs_service gender  salary
0      Prof          B       19           18   Male  139750
1      Prof          B       20           16   Male  173200
2  AsstProf          B        4            3   Male   79750
3      Prof          B       45           39   Male  115000
4      Prof          B       40           41   Male  141500

In [ ]:
df = df[['rank','yrs_service','salary']]

In [ ]:
rank_dummy = pd.get_dummies(df['rank'],drop_first=True)

#creating dummy columns

In [ ]:
rank_dummy.head()

#no of dummy columns = no of categories - 1 -----> 3-1=2

AsstProf  Prof
0         0     1
1         0     1
2         1     0
3         0     1
4         0     1

In [ ]:
df= df.drop('rank',axis=1)

In [ ]:
df = pd.concat([df,rank_dummy],axis=1)

In [ ]:
df.head()

yrs_service  salary  AsstProf  Prof
0           18  139750         0     1
1           16  173200         0     1
2            3   79750         1     0
3           39  115000         0     1
4           41  141500         0     1

In [ ]:
x_train,x_test,y_train,y_test = train_test_split(df.drop('salary',axis=1),df['salary'],test_size=0.25)

#splitting the dataset

In [ ]:
x_train = sm.add_constant(x_train,prepend=False)

#adding constant

In [ ]:
mod1 = sm.OLS(y_train,x_train).fit()

#creating model

In [ ]:
print(mod1.summary())

#summary

                            OLS Regression Results                            
Dep. Variable:                 salary   R-squared:                       0.387
Model:                            OLS   Adj. R-squared:                  0.381
Method:                 Least Squares   F-statistic:                     61.69
Date:                Tue, 20 Jun 2023   Prob (F-statistic):           6.09e-31
Time:                        11:22:05   Log-Likelihood:                -3409.7
No. Observations:                 297   AIC:                             6827.
Df Residuals:                     293   BIC:                             6842.
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
                  coef    std err          t      P>|t|      [0.025      0.975]
-------------------------------------------------------------------------------
yrs_service  -205.0711    132.152     -1.552      

yrs_service coeff is -ve, -ve correlation, inc in yrs of ser, salary will come down.
yrs service is continuus data.
it has -ve correlation with salary.
as p>0.05 , H0 is not rejected.

**H0: yrs of service is not impact on salary**

rank= assosciate prof, assistant prof, prof

assocate prof is taken as a reference event.

**H0 for assistant prof**:
--> **if anyone moves from assosiate prof to assitant prof salary remains the same or the movement of  associate prof to assistant prof will not influence salary**.
OR
--> **if anyone moves from assosiate prof to assitant prof then the change in salary is statistically insignificant**.




--> **coeff of assistant prof represents that, when you move from assosiate to assist prof then salary will come down by -1.537+e04 by keeping the yrs of service constant**.g

OR

--> **coeff of assistant prof represents that, when you become assist prof with respect assosiate prof, then salary will come down by -1.537+e04 by keeping the yrs of service constant**.
--> the change is significant beecause p<0.05


--> **coeff of prof represents that,  when you move from assosiate to aprof then salary will increase by 3.413e+04 by keeping the yrs of service constant**.
--> as p<0.05 , the change is significant.

--> prof has more impact on salary than assistant prof as p(prof) < p(assistant prof).

--> T**he constant represents the salary of associate prof salary when years of service=0, coeff of asst prof = 0, coeff of prof = 0**
that means, it is the saalry of associate prof who is a fresher.


--> REFER your notes for this example continuation for clear understanding.

y = mx + c (Single Linear Regression)
y = m1x1 + m2x2 + c  (Multiple Liner Regression)

c1 = Asst Prof
c2 = Prof
sal = yrs*(-195) - 161440*C1 + 3421*c2 + 9601
if yrs od exp=3, Asst Prof = 3*(-195)-161440+0+9601
0, Asst Prof = 0+0+0+9601

**Logistic Regression**:

--> The predictors can be binary, categorical (with multiple categories), ordinal, or continuous variables
--> Response is categorical

**Types of Logistic regression**:-

 **1. BINARY LOGISTIC REGRESSION**

--> when response is a binary data (Y/N)
 Binary Logistic Regression
-> predict when someone get job or not OR buy my insurance or not

**2. NOMINAL LOGISTIC REGRESSION**

response -> categorical
order is not important
atleast 3 options honge


**3. ORDINAL LOGISTIC REGRESSION**

--> response - categorical
--> categories -> ordered